jerry-core/ecma/base/ecma-helpers-value.c

```diff
@@ -275,21 +300,156 @@ ecma_check_value_type_is_spec_defined (ecma_value_t value) /**< ecma value */
 {
-  JERRY_ASSERT (ecma_num_p != NULL);
+  JERRY_ASSERT (ECMA_INTEGER_NUMBER_MIN <= integer_value
+                && integer_value <= ECMA_INTEGER_NUMBER_MAX);
 
-  return ecma_pointer_to_ecma_value (ecma_num_p) | ECMA_TYPE_NUMBER;
+  return (ecma_value_t) ((integer_value << ECMA_DIRECT_SHIFT) | ECMA_TYPE_DIRECT);
+} /* ecma_make_integer_value */
+
+/**
+ * Allocate and initialize a new float number without checks.
+ *
+ * @return ecma-value
+ */
+static ecma_value_t __attr_const___
+ecma_create_float_number (ecma_number_t ecma_num) /**< value of the float number */
+{
+  ecma_number_t *ecma_num_p = ecma_alloc_number ();
+
+  *ecma_num_p = ecma_num;
+
+  return ecma_pointer_to_ecma_value (ecma_num_p) | ECMA_TYPE_FLOAT;
+} /* ecma_create_float_number */
+
+/**
+ * Create a new float number.
+ *
+ * This function is intended to create floating point constants when it is
+ * known that ecma_num is not an integer number, e.g. 0.5, NaN, or -infinity.
+ * Calling this function is faster than ecma_make_number_value since no
+ * checks are performed on the input apart from a debug assert.
+ *
+ * @return ecma-value
+ */
+inline ecma_value_t __attr_always_inline___
+ecma_make_float_number_value (ecma_number_t ecma_num) /**< value of the float number */
+{
+  /* Simple check to ensure that ecma_num cannot be converted to integer. */
+  JERRY_ASSERT (((ecma_number_t) (ecma_integer_value_t) ecma_num) != ecma_num);
+
+  return ecma_create_float_number (ecma_num);
+} /* ecma_make_float_number_value */
+
+/**
+ * Checks whether the passed number is +0.0
+ *
+ * @return true, if it is +0.0, false otherwise
+ */
+static inline bool __attr_const___ __attr_always_inline___
+ecma_is_number_equals_to_positive_zero (ecma_number_t ecma_num) /**< number */
+{
+#if CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT32
+  union
+  {
+    uint32_t u32_value;
+    ecma_number_t float_value;
+  } u;
+
+  u.float_value = ecma_num;
+
+  return u.u32_value == 0;
+#else /* CONFIG_ECMA_NUMBER_TYPE != CONFIG_ECMA_NUMBER_FLOAT32 */
+  union
+  {
+    uint64_t u64_value;
+    ecma_number_t float_value;
+  } u;
+
+  u.float_value = ecma_num;
+
+  return u.u64_value == 0;
+#endif /* CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT32 */
+} /* ecma_is_number_equals_to_positive_zero */
+
+/**
+ * Encode a number into an ecma-value
+ *
+ * @return ecma-value
+ */
+ecma_value_t
+ecma_make_number_value (ecma_number_t ecma_num) /**< number to be encoded */
+{
+  ecma_integer_value_t integer_value = (ecma_integer_value_t) ecma_num;
+
+  if ((ecma_number_t) integer_value == ecma_num)
+  {
+    if (integer_value == 0)
+    {
+      if (ecma_is_number_equals_to_positive_zero (ecma_num))
+      {
+        return ecma_make_integer_value (0);
+      }
+    }
+    else if (ECMA_INTEGER_NUMBER_MIN <= integer_value
+             && integer_value <= ECMA_INTEGER_NUMBER_MAX)
+    {
+      return ecma_make_integer_value (integer_value);
+    }
+  }
```

Comment:
The above is incorrect, because the ECMA_INTEGER_NUMBER_MIN <= integer_value && integer_value <= ECMA_INTEGER_NUMBER_MAX part includes the integer_value == 0

That is, it is the same as:
```
if (((ecma_number_t) integer_value == ecma_num)
    && (ECMA_INTEGER_NUMBER_MIN <= integer_value && integer_value <= ECMA_INTEGER_NUMBER_MAX)))
```
probably a ?: is better